### ISSUE TITLE
Improve shrink-to-fit layout (fit-content sizing)

### ISSUE THREAD

**Issue by amlinux** [mikke89/RmlUi#552]
# Absolutely positioned `div`

```
<rml>
	<head>
		<style>
			* {
				box-sizing: border-box;
		    	        font-family: Play;
			}

			.block1 {
			        display: block;
				border: 1px red;
				width: 100px;
			}

			.block2 {
			        display: block;
				border: 1px blue;
				width: 100px;
			}

			.block3 {
			        display: block;
				border: 1px green;
				width: 100px;
			}

			#wrapper {
				position: absolute;
				display: block;
				left: 100px;
				top: 100px;
				border: 1px white;
			}
		</style>
	</head>
	<body>
		<div id="wrapper">
			<div class="block1">Block 1</div>
			<div class="block2">Block 2</div>
			<div class="block3">Block 3</div>
		</div>
	</body>
</rml>
```

Note that white `#wrapper` that was supposed to wrap all nested blocks, in fact has width = 2 (empty content + 2 pixels for borders):

![image](https://github.com/mikke89/RmlUi/assets/449322/c222be5c-ec04-472e-91bd-4d44e0465c09)

If I remove "position: absolute", it does the right thing:

![image](https://github.com/mikke89/RmlUi/assets/449322/61737761-6919-478f-9612-4c2408273726)

# Blocks with `display: flex`

```
<rml>
	<head>
		<style>
			* {
				box-sizing: border-box;
		    	        font-family: Play;
			}

			.outer {
			        display: flex;
			        border: 1px red;
                                padding: 50px;
			}

                       .inner {
                                border: 1px blue;
                                padding: 50px;
                       }
		</style>
	</head>
	<body>
                <div class="outer"><div class="inner">Block 1</div></div>
	</body>
</rml>
```

Note that outer (red) block doesn't "see" that its inner (blue) block has any width:

![image](https://github.com/mikke89/RmlUi/assets/449322/163feaf0-5d91-46df-9ca4-7a81b3be379e)

# Complex grid layouts

```
<rml>
    <head>
        <style>
            * {
                box-sizing: border-box;
                font-family: Play;
            }

            .block {
                display: block;
                width: 50px;
                height: 50px;
                margin: 10px;
            }

            .red {
                background-color: red;
            }

            .green {
                background-color: green;
            }

            .blue {
                background-color: blue;
            }

            .outer {
                display: flex;
                flex-direction: row;
            }

            .inner {
                display: flex;
                flex-direction: column;
            }
        </style>
    </head>
    <body>
        <div class="outer">
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
        </div>
    </body>
</rml>
```

This is how RmlUi renders it:

![image](https://github.com/mikke89/RmlUi/assets/449322/aa05b52d-cd95-4a43-b733-5b45ea6ad5a5)

Equivalent layout in HTML/CSS (expected result):

![image](https://github.com/mikke89/RmlUi/assets/449322/795a4f3f-809c-4803-820d-566e795b2d52)

Interesting is that if I swap `flex-direction` in outer and inner blocks (outer to `column`, inner  to `row`), then RmlUi doesn't render anything:

![image](https://github.com/mikke89/RmlUi/assets/449322/fde21a2e-07ab-4e11-9854-76d1fa4830ce)

While HTML/CSS renders it as expected:

![image](https://github.com/mikke89/RmlUi/assets/449322/73362107-0ff2-42e2-9885-30e5da07505a)

I believe that the root cause is the same as in https://github.com/mikke89/RmlUi/issues/520, and while https://github.com/mikke89/RmlUi/commit/8a94ab779b359f3fbcea5bf0cbc2d3373bdba179 fixed it for a trivial case, a larger body of cases is still not supported.

**Comment by mikke89** [comment]
Thanks for the detailed report, and the reproducible cases. This is certainly an area of the layout engine that can be improved, that I would fully support. These situations are all related to calculations of *shrink-to-fit width* as its called in CSS2, or more generically [*fit-content size* in CSS3](https://drafts.csswg.org/css-sizing/#fit-content-size).

Currently, this is done quite ad-hoc within the layout engine, and not supported for tables and flexboxes in particular.

Some notes for implementing this:
- We need to implement some notion of *max-content* sizing, as seen in the definition of fit-content size in the above link.
- In my view, the *min-content* sizing from CSS is not particularly important for us, since content in RmlUi layouts are considered to be better controlled than on the web in general. I.e. we explicitly don't try to "fix" broken layouts.
- In particular, for flexbox, we need to implement [intrinsic sizing of flex containers](https://drafts.csswg.org/css-flexbox/#intrinsic-sizes). For tables, see [max-content here](https://www.w3.org/TR/css-tables-3/#max-content-width-of-a-table).
- We could need a more robust approach to determine intrinsic sizing in the layout engine. We may want some state that keeps track of any max-content constraint, the current available width, definite vs indefinite sizes, and similar. This could also help guide line breaks and early-outs for performance reasons.
- Currently, we only consider width direction, we may also need a notion of fit-content height, in particular when working with flexboxes with column direction.

Performance considerations:
- Each layout element that invokes "shrink-to-fit" essentially has to do layout twice. Once to determine the max-content width, and once more for the actually determined width. 
- For nested elements that invoke shrink-to-fit, this essentially means the number of layout operations for the most-descendant elements grows by `O(n^2)`, where `n` is the number of nested shrink-to-fit elements.
- If we implement *min-content* as well, then that expands to doing layout three times per shrink-to-fit element.
- I'm sure we can be much smarter about this, such that these calculations can either break early under certain situations, or some ways of avoiding to do the layout multiple times in some situations. However, in general this sounds like it would increase complexity, require a lot of testing, and probably a lot of development efforts. But there could be some low-hanging fruits.
  - In particular, we could break out early if the max-content width exceeds the available width under a fit-content constraint.

Hopefully, this gives some notion of how we could get started. I'm sure this is something that needs a lot of iterations to solve more generally, we'll have to test it against a lot of different situations as it evolves. But these examples seem like good starting points.

**Comment by mikke89** [comment]
This issue should be resolved with #559 being merged.

More generally of course, there are always improvements that are possible in terms of shrink-to-fit, but any specific situations should be discussed in new issues.
